src/raft_server.c

```diff
@@ -4915,131 +4915,188 @@ raft_server_net_client_request_init_sm_apply(
 void
 raft_server_backend_setup_last_applied(struct raft_instance *ri,
                                        raft_entry_idx_t last_applied_idx,
+                                       raft_entry_idx_t last_applied_sub_idx,
+                                       raft_entry_idx_t last_applied_sub_idx_max,
                                        crc32_t last_applied_cumulative_crc)
 {
     NIOVA_ASSERT(ri && (raft_instance_is_booting(ri) ||
                         raft_instance_is_recovering(ri)));
 
     ri->ri_last_applied_idx = last_applied_idx;
     ri->ri_last_applied_synced_idx = last_applied_idx;
+    ri->ri_last_applied_sub_idx = last_applied_sub_idx;
+    ri->ri_last_applied_sub_idx_max = last_applied_sub_idx_max;
     ri->ri_last_applied_cumulative_crc = last_applied_cumulative_crc;
 
+
     DBG_RAFT_INSTANCE(LL_TRACE, ri, "");
 }
 
 static raft_server_epoll_sm_apply_t
 raft_server_last_applied_increment(struct raft_instance *ri,
                                    const struct raft_entry_header *reh)
 {
-    NIOVA_ASSERT(ri && reh &&
-                 (reh->reh_index == (ri->ri_last_applied_idx + 1)));
+    NIOVA_ASSERT(ri && reh && 
+        (reh->reh_num_entries > 0));
+    
+    /*
+    Increment the sub apply idx, if its the first sub entry then increment
+    the main last applied idx and update the cumulative crc. If the sub idx
+    reaches the max sub idx then reset both sub idx and max sub idx to -1.
+    */
+    ri->ri_last_applied_sub_idx++;
+
+    if (ri->ri_last_applied_sub_idx == 0) {
+        ri->ri_last_applied_idx++;
+        ri->ri_last_applied_cumulative_crc ^= reh->reh_crc;
+        ri->ri_last_applied_sub_idx_max = reh->reh_num_entries - 1;
+    }
+
+    if (ri->ri_last_applied_sub_idx == ri->ri_last_applied_sub_idx_max) {
+        ri->ri_last_applied_sub_idx = -1;
+        ri->ri_last_applied_sub_idx_max = -1;
+    }
 
-    ri->ri_last_applied_idx++;
-    ri->ri_last_applied_cumulative_crc ^= reh->reh_crc;
+    DBG_RAFT_INSTANCE(LL_NOTIFY, ri, "idx=%ld crc=%x sub-idx=%ld/%ld",
+                      ri->ri_last_applied_idx, reh->reh_crc,
+                      ri->ri_last_applied_sub_idx, ri->ri_last_applied_sub_idx_max);
+}
 
-    DBG_RAFT_INSTANCE(LL_NOTIFY, ri, "idx=%ld crc=%x",
-                      ri->ri_last_applied_idx, reh->reh_crc);
+static bool
+raft_server_needs_apply(struct raft_instance *ri) {
```

Comment:
I believe `ri` can be `const`.  Whenever possible we should applying `const` to input parameters.